test/integration/base.py

```diff
@@ -1,50 +1,46 @@
-import unittest
-
-from .pelikan_client import PelikanClient
-from .pelikan_server import PelikanServer
-
-SERVER_PORT = 12345
-ADMIN_PORT = 12346
+from client import DataClient, AdminClient
+from loader import load_seq
+from server import PelikanServer
 
+import unittest
 
-class PelikanTest(unittest.TestCase):
-    reserved_ports = set()
+DEFAULT_SERVER = ('localhost', 12321)
+DEFAULT_ADMIN = ('localhost', 9999)
 
+class GenericTest(unittest.TestCase):
     def setUp(self):
-        self.server = self.getServer(
-            server_port=SERVER_PORT,
-            admin_port=ADMIN_PORT
-        )
-        self.client = PelikanClient(
-            server_port=SERVER_PORT,
-            admin_port=ADMIN_PORT
-        )
+        self.server = PelikanServer('pelikan_twemcache')
+        self.server.ready()
+        self.data_client = DataClient(DEFAULT_SERVER)
+        self.admin_client = AdminClient(DEFAULT_ADMIN)
+        self.stats = self.admin_client.stats()
 
     def tearDown(self):
+        self.data_client.close()
+        self.admin_client.close()
         self.server.stop()
 
-    def assertRead(self, expected):
-        read = self.client.read(len(expected))
-        self.assertEqual(expected, read)
-
-
-    def assertMetrics(self, *args):
-        stats = self.client.getStats()
-        for (k, v) in args:
-            self.assertEqual(
-                stats.get(k, None),
-                str(v),
-                'Expected {} to be {}, got {} instead'.format(
-                    k,
-                    v,
-                    stats.get(k, None),
-                )
-            )
-
-
-class TwemcacheTest(PelikanTest):
-    def getServer(self, **kwargs):
-        server = PelikanServer('pelikan_twemcache')
-        server.start(**kwargs)
-        server.wait_ready()
-        return server
+    def load(self, fname):
+        """loading a test sequence from a file"""
+        self.seq = load_seq(fname)
+
+    def assertResponse(self, expected):
+        """receive and verify response (a list) matches expectation"""
+        if len(expected) > 0:
+            rsp = self.data_client.response()
+            self.assertEqual(len(rsp), len(expected))
+            for i in range(len(rsp)):
+                self.assertEqual(rsp[i], expected[i])
+
+    def assertStats(self, delta):
+        """delta, a dict, captures the expected change in a subset of metrics"""
+        stats = self.admin_client.stats()
+        for k in delta:
+            self.assertEqual(int(stats[k]) - int(self.stats[k]), delta[k])
```

Comment:
This could use a third parameter (an error message) including `k`, otherwise getting `3 is not equal to 2` is hard to understand, specially since many parameters may be checked in the same stacktrace.